### PR TITLE
Differentiate between API errors and application exit codes.

### PR DESCRIPTION
**Pull request by BillyONeal**
This is my competing resolution of #483 . Rather than introducing a new type, it uses ExpectedL.

Fixes https://github.com/microsoft/vcpkg/issues/21800

### CODE REVIEW

File: include/vcpkg/base/expected.h [hunk 1]
```diff
@@ -14,6 +14,26 @@
 
 namespace vcpkg
 {
+    struct Unit
+    {
+        // A meaningless type intended to be used with Expected when there is no meaningful value.
+    };
```
Comment (autoantwort) [inline include/vcpkg/base/expected.h:20]: Why no `Expected<void>`?

File: include/vcpkg/base/system.process.h [hunk 2]
```diff
@@ -150,4 +152,12 @@ namespace vcpkg
     void enter_interactive_subprocess();
     void exit_interactive_subprocess();
 #endif
+
+    // If exit code is 0, returns a 'success' ExpectedS.
+    // Otherwise, returns an ExpectedS containing error text
+    ExpectedS<Unit> flatten(const ExpectedApi<ExitCodeAndOutput>&, StringView tool_name);
```
Comment (autoantwort) [inline include/vcpkg/base/system.process.h:158]: Why not `ExpectedL`?

File: include/vcpkg/base/expected.h (hunk 1 above)
Comment (BillyONeal) [reply include/vcpkg/base/expected.h]: We decided to not do the "duplicate all the code in expected" solution in https://github.com/microsoft/vcpkg-tool/pull/564

File: include/vcpkg/base/system.process.h [hunk 3]
```diff
@@ -119,28 +120,29 @@ namespace vcpkg
...
-                                                     EchoInDebug echo_in_debug = EchoInDebug::Hide);
+    ExpectedApi<ExitCodeAndOutput> cmd_execute_and_capture_output(
+        const Command& cmd_line,
+        const WorkingDirectory& wd = default_working_directory,
+        const Environment& env = default_environment,
+        Encoding encoding = Encoding::Utf8,
+        EchoInDebug echo_in_debug = EchoInDebug::Hide);
 
-    std::vector<ExitCodeAndOutput> cmd_execute_and_capture_output_parallel(
+    std::vector<ExpectedApi<ExitCodeAndOutput>> cmd_execute_and_capture_output_parallel(
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply include/vcpkg/base/system.process.h]: I decided that I would prefer to land this as-is on this point but would not object to making a change like that in the future.

File: include/vcpkg/base/system.process.h (hunk 2 above)
Comment (BillyONeal) [reply include/vcpkg/base/system.process.h]: Done!

**Review by ras0219-msft — APPROVED** [review]

File: include/vcpkg/base/system.process.h [hunk 4]
```diff
@@ -150,4 +151,14 @@ namespace vcpkg
...
+
+    bool succeeded(const ExpectedL<int>& maybe_exit) noexcept;
+
+    // If exit code is 0, returns a 'success' ExpectedL.
+    // Otherwise, returns an ExpectedL containing error text
+    ExpectedL<Unit> flatten(const ExpectedL<ExitCodeAndOutput>&, StringView tool_name);
+
+    // If exit code is 0, returns a 'success' ExpectedL containing the output
+    // Otherwise, returns an ExpectedL containing error text
+    ExpectedL<std::string> flatten_out(const ExpectedL<ExitCodeAndOutput>&, StringView tool_name);
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply include/vcpkg/base/system.process.h]: Done!

File: src/vcpkg/base/system.process.cpp [hunk 5]
```diff
@@ -932,4 +995,56 @@ namespace vcpkg
...
+                    .append_raw(maybe_exit.error().to_string())};
+    }
+
+    ExpectedL<std::string> flatten_out(const ExpectedL<ExitCodeAndOutput>& maybe_exit, StringView tool_name)
+    {
+        if (auto exit = maybe_exit.get())
+        {
+            if (exit->exit_code == 0)
+            {
+                return {exit->output, expected_left_tag};
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/base/system.process.cpp]: Done!

**Review by vicroms — APPROVED** [review]
I have some minor nits about magic numbers like in: 

```
if (i < 19 && code_and_output->exit_code == 1618)

// or 

 out->push_back(std::strtol(line.data() + guid_marker.size(), nullptr, 10));
```

but most can't be understood based on context (and they were already present in the code before your changes) so it's OK

**Pull request merged**